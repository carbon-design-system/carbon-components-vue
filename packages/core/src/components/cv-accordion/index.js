import CvAccordion from './cv-accordion';
import CvAccordionItem from './cv-accordion-item';
import CvAccordionSkeleton from './cv-accordion-skeleton';

export { CvAccordion, CvAccordionItem, CvAccordionSkeleton };
