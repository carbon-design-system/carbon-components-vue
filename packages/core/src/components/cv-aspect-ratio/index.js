import CvAspectRatio from './cv-aspect-ratio';
export { CvAspectRatio };
