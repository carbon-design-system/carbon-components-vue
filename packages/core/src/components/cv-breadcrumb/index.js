import CvBreadcrumb from './cv-breadcrumb';
import CvBreadcrumbItem from './cv-breadcrumb-item';
import CvBreadcrumbSkeleton from './cv-breadcrumb-skeleton';

export { CvBreadcrumb, CvBreadcrumbItem, CvBreadcrumbSkeleton };
