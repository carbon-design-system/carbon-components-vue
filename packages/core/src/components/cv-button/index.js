import CvButton from './cv-button';
import CvIconButton from './cv-icon-button';
import CvButtonSkeleton from './cv-button-skeleton';
import CvButtonSet from './cv-button-set';

export { CvButton, CvIconButton, CvButtonSkeleton, CvButtonSet };
