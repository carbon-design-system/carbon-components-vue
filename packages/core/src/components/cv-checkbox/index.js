import CvCheckbox from './cv-checkbox';
import CvCheckboxSkeleton from './cv-checkbox-skeleton';
export { CvCheckbox, CvCheckboxSkeleton };
