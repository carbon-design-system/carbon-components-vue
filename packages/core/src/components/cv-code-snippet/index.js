import CvCodeSnippet from './cv-code-snippet';
import CvCodeSnippetSkeleton from './cv-code-snippet-skeleton';
export { CvCodeSnippet, CvCodeSnippetSkeleton };
