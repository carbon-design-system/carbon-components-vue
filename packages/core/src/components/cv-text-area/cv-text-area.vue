<template>
  <div :class="`cv-text-area ${carbonPrefix}--form-item`">
    <label
      :for="uid"
      :class="[
        `${carbonPrefix}--label`,
        {
          [`${carbonPrefix}--label--disabled`]: $attrs.disabled !== undefined && $attrs.disabled,
        },
      ]"
      >{{ label }}</label
    >
    <div :class="`${carbonPrefix}--text-area__wrapper`" :data-invalid="isInvalid">
      <WarningFilled16 v-if="isInvalid" :class="`${carbonPrefix}--text-area__invalid-icon`" />
      <textarea
        :id="uid"
        :class="[
          `${carbonPrefix}--text-area`,
          {
            [`${carbonPrefix}--text-area--light`]: theme === 'light',
            [`${carbonPrefix}--text-area--invalid`]: isInvalid,
          },
        ]"
        v-bind="$attrs"
        :value="value"
        v-on="inputListeners"
        ref="textarea"
      ></textarea>
    </div>
    <div :class="`${carbonPrefix}--form-requirement`" v-if="isInvalid">
      <slot name="invalid-message">{{ invalidMessage }}</slot>
    </div>
    <div
      v-if="!isInvalid && isHelper"
      :class="[
        `${carbonPrefix}--form__helper-text`,
        { [`${carbonPrefix}--form__helper-text--disabled`]: $attrs.disabled },
      ]"
    >
      <slot name="helper-text">{{ helperText }}</slot>
    </div>
  </div>
</template>

<script>
import { uidMixin, themeMixin, methodsMixin, carbonPrefixMixin } from '../../mixins';
import WarningFilled16 from '@carbon/icons-vue/es/warning--filled/16';

export default {
  name: 'CvTextArea',
  mixins: [uidMixin, themeMixin, carbonPrefixMixin, methodsMixin({ textarea: ['blur', 'focus'] })],
  inheritAttrs: false,
  components: { WarningFilled16 },
  props: {
    helperText: { type: String, default: undefined },
    invalidMessage: { type: String, default: undefined },
    label: String,
    value: String,
  },
  data() {
    return {
      isHelper: false,
      isInvalid: false,
    };
  },
  mounted() {
    this.checkSlots();
  },
  updated() {
    this.checkSlots();
  },
  computed: {
    // Bind listeners at the component level to the embedded input element and
    // add our own input listener to service the v-model. See:
    // https://vuejs.org/v2/guide/components-custom-events.html#Customizing-Component-v-model
    inputListeners() {
      return {
        ...this.$listeners,
        input: event => this.$emit('input', event.target.value),
      };
    },
  },
  methods: {
    checkSlots() {
      // NOTE: this.$slots is not reactive so needs to be managed on updated
      this.isInvalid = !!(this.$slots['invalid-message'] || (this.invalidMessage && this.invalidMessage.length));
      this.isHelper = !!(this.$slots['helper-text'] || (this.helperText && this.helperText.length));
    },
  },
};
</script>
