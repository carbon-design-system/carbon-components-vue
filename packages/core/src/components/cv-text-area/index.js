import CvTextArea from './cv-text-area';
export { CvTextArea };
