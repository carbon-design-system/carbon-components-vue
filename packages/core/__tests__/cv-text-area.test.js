// import { shallowMount as shallow, mount } from '@vue/test-utils';
import { testComponent, awaitNextTick } from './_helpers';
const { shallowMount: shallow, mount, trigger } = awaitNextTick;
import { CvTextArea } from '@/components/cv-text-area';

describe('CvTextArea', () => {
  // ***************
  // PROP CHECKS
  // ***************
  testComponent.propsAreType(CvTextArea, ['helperText', 'invalidMessage', 'label', 'value', 'id'], String);
  testComponent.propsHaveDefaultOfUndefined(CvTextArea, ['helperText', 'invalidMessage', 'light']);

  // ***************
  // SNAPSHOT TESTS
  // ***************
  it('should render correctly', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1' };
    const wrapper = await shallow(CvTextArea, { propsData });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when light theme is used', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1', theme: 'light' };
    const wrapper = await shallow(CvTextArea, { propsData });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when invalid message is provided', async () => {
    const propsData = { label: 'test label', value: 'test value', invalidMessage: 'invalid test message', id: '1' };
    const wrapper = await shallow(CvTextArea, { propsData });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when invalid message slot is provided', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1' };
    const wrapper = await shallow(CvTextArea, {
      slots: {
        'invalid-message': '<div class="invalid-message-class">invalid message slot</div>',
      },
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when helper text is provided', async () => {
    const propsData = { label: 'test label', value: 'test value', helperText: 'helper test message', id: '1' };
    const wrapper = await shallow(CvTextArea, { propsData });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render correctly when helper text slot is provided', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1' };
    const wrapper = await shallow(CvTextArea, {
      slots: {
        'helper-text': '<div class="helper-text-class">helper text slot</div>',
      },
      propsData,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  // ***************
  // FUNCTIONAL TESTS
  // ***************

  it('inValid should be true when invalid message is provided', async () => {
    const propsData = { label: 'test label', value: 'test value', invalidMessage: 'invalid test message', id: '1' };
    const wrapper = await shallow(CvTextArea, { propsData });
    expect(wrapper.vm.isInvalid).toEqual(true);
  });

  it('inValid should be true when invalid message slot is provided', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1' };
    const wrapper = await mount(CvTextArea, {
      slots: {
        'invalid-message': '<div class="invalid-message-class">invalid message slot</div>',
      },
      propsData,
    });
    expect(wrapper.vm.isInvalid).toEqual(true);
  });

  it('inValid should be false when invalid message slot and invalid message are not provided', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1' };
    const wrapper = await shallow(CvTextArea, { propsData });
    expect(wrapper.vm.isInvalid).toEqual(false);
  });

  it('isHelper should be true when helper text is provided', async () => {
    const propsData = { label: 'test label', value: 'test value', helperText: 'helper text test message', id: '1' };
    const wrapper = await shallow(CvTextArea, { propsData });
    expect(wrapper.vm.isHelper).toEqual(true);
  });

  it('isHelper should be true when helper text slot is provided', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1' };
    const wrapper = await shallow(CvTextArea, {
      slots: {
        'helper-text': '<div class="helper-text-class">helper text slot</div>',
      },
      propsData,
    });
    expect(wrapper.vm.isHelper).toEqual(true);
  });

  it('isHelper should be false when helper text slot and helper text are not provided', async () => {
    const propsData = { label: 'test label', value: 'test value', id: '1' };
    const wrapper = await shallow(CvTextArea, { propsData });
    expect(wrapper.vm.isHelper).toEqual(false);
  });

  it('should emit current value on change', async () => {
    const value = 'test value';
    const propsData = { label: 'test label', value, id: '1' };
    const wrapper = await shallow(CvTextArea, { propsData });
    await trigger(wrapper.find('textarea'), 'input');
    expect(wrapper.emitted().input[0]).toEqual([value]);
  });
});
