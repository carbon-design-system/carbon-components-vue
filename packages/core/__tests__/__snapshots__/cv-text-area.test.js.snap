// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CvTextArea should render correctly 1`] = `
<div class="cv-text-area bx--form-item"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-area__wrapper">
    <!----> <textarea id="1" class="bx--text-area"></textarea></div>
  <!---->
  <!---->
</div>
`;

exports[`CvTextArea should render correctly when helper text is provided 1`] = `
<div class="cv-text-area bx--form-item"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-area__wrapper">
    <!----> <textarea id="1" class="bx--text-area"></textarea></div>
  <!---->
  <div class="bx--form__helper-text">helper test message</div>
</div>
`;

exports[`CvTextArea should render correctly when helper text slot is provided 1`] = `
<div class="cv-text-area bx--form-item"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-area__wrapper">
    <!----> <textarea id="1" class="bx--text-area"></textarea></div>
  <!---->
  <div class="bx--form__helper-text">
    <div class="helper-text-class">helper text slot</div>
  </div>
</div>
`;

exports[`CvTextArea should render correctly when invalid message is provided 1`] = `
<div class="cv-text-area bx--form-item"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-area__wrapper" data-invalid="true">
    <warningfilled16-stub class="bx--text-area__invalid-icon"></warningfilled16-stub> <textarea id="1" class="bx--text-area bx--text-area--invalid" aria-invalid="true" aria-describedby="error-1"></textarea>
  </div>
  <div id="error-1" class="bx--form-requirement">invalid test message</div>
  <!---->
</div>
`;

exports[`CvTextArea should render correctly when invalid message slot is provided 1`] = `
<div class="cv-text-area bx--form-item"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-area__wrapper" data-invalid="true">
    <warningfilled16-stub class="bx--text-area__invalid-icon"></warningfilled16-stub> <textarea id="1" class="bx--text-area bx--text-area--invalid" aria-invalid="true" aria-describedby="error-1"></textarea>
  </div>
  <div id="error-1" class="bx--form-requirement">
    <div class="invalid-message-class">invalid message slot</div>
  </div>
  <!---->
</div>
`;

exports[`CvTextArea should render correctly when light theme is used 1`] = `
<div class="cv-text-area bx--form-item"><label for="1" class="bx--label">test label</label>
  <div class="bx--text-area__wrapper">
    <!----> <textarea id="1" class="bx--text-area bx--text-area--light"></textarea></div>
  <!---->
  <!---->
</div>
`;
